sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteGetJsonObject.scala

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Expression, GetJsonObject, JsonTuple, NamedExpression}
+import org.apache.spark.sql.catalyst.plans.logical.{Generate, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.GET_JSON_OBJECT
+import org.apache.spark.sql.types.StringType
+
+/**
+ * This rule rewrites multiple GetJsonObjects to a JsonTuple if their json expression is the same.
+ */
+object RewriteGetJsonObject extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUpWithPruning(
+    _.containsPattern(GET_JSON_OBJECT), ruleId) {
+    case p: Project =>
+      val getJsonObjects = p.projectList.flatMap {
+        _.collect {
+          case gjo: GetJsonObject if gjo.rewriteToJsonTuplePath.nonEmpty => gjo
+        }
+      }
+
+      val groupedGetJsonObjects = getJsonObjects.groupBy(_.json).filter(_._2.size > 1)
+      if (groupedGetJsonObjects.nonEmpty) {
+        var newChild = p.child
+        val keyValues = mutable.LinkedHashMap.empty[Expression, AttributeReference]
+        groupedGetJsonObjects.foreach {
+          case (json, getJsonObjects) =>
+            val generatorOutput = getJsonObjects.map { j =>
+              val attr = AttributeReference(j.rewriteToJsonTuplePath.get.toString, StringType)()
+              keyValues.put(j.canonicalized, attr)
+              attr
+            }
+            newChild = Generate(
```

Comment:
are you sure the extra `Generate` won't slow down the query?

Reply:
shall we add a scalar version of `JsonTuple`?

Reply:
or add a function to turn json string into a row, so that extracting the json field is very cheap?

Reply:
> are you sure the extra Generate won't slow down the query?

GetJsonObject supports codegen but JsonTuple does not, so the transformation may slow down the query by breaking the WSCG in certain cases.